x/ibc/03-connection/keeper/handshake_test.go

```diff
@@ -1,137 +1,161 @@
+			// chainA connection not created
 		}, false},
-		{"invalid previous connection", func() uint64 {
-			suite.chainB.CreateClient(suite.chainA)
-			suite.chainA.CreateClient(suite.chainB)
-			suite.chainB.createConnection(testConnectionIDB, testConnectionIDA, testClientIDA, testClientIDB, types.UNINITIALIZED)
-			suite.chainB.updateClient(suite.chainA)
-			suite.chainA.updateClient(suite.chainB)
-			return 0
+		{"consensus state verification failed", func() {
+			clientA, clientB = suite.coordinator.SetupClients(suite.chainA, suite.chainB, clientexported.Tendermint)
+			// give chainA wrong consensus state for chainB
+			consState, found := suite.chainA.App.IBCKeeper.ClientKeeper.GetLatestClientConsensusState(suite.chainA.GetContext(), clientA)
+			suite.Require().True(found)
+
+			tmConsState, ok := consState.(ibctmtypes.ConsensusState)
+			suite.Require().True(ok)
+
+			tmConsState.Timestamp = time.Now()
+			suite.chainA.App.IBCKeeper.ClientKeeper.SetClientConsensusState(suite.chainA.GetContext(), clientA, tmConsState.Height, tmConsState)
+
+			_, _, err := suite.coordinator.ConnOpenInit(suite.chainA, suite.chainB, clientA, clientB)
+			suite.Require().NoError(err)
 		}, false},
-		{"couldn't add connection to client", func() uint64 {
-			suite.chainB.CreateClient(suite.chainA)
-			suite.chainA.createConnection(testConnectionIDB, testConnectionIDA, testClientIDB, testClientIDA, types.UNINITIALIZED)
-			suite.chainB.updateClient(suite.chainA)
-			suite.chainA.updateClient(suite.chainB)
-			return 0
+		{"invalid previous connection is in TRYOPEN", func() {
+			clientA, clientB = suite.coordinator.SetupClients(suite.chainA, suite.chainB, clientexported.Tendermint)
+
+			// open init chainA
+			connA, connB, err := suite.coordinator.ConnOpenInit(suite.chainA, suite.chainB, clientA, clientB)
+			suite.Require().NoError(err)
+
+			// open try chainB
+			err = suite.coordinator.ConnOpenTry(suite.chainB, suite.chainA, connB, connA)
+			suite.Require().NoError(err)
 		}, false},
 	}
 
-	for i, tc := range testCases {
+	for _, tc := range testCases {
 		tc := tc
-		i := i
-		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
-			suite.SetupTest() // reset
 
-			consensusHeight := tc.malleate()
+		suite.Run(tc.msg, func() {
+			suite.SetupTest()       // reset
+			consensusHeightDiff = 0 // must be explicity changed in malleate
+
+			tc.malleate()
+
+			connA := suite.chainA.GetFirstTestConnection(clientA, clientB)
+			connB := suite.chainB.GetFirstTestConnection(clientB, clientA)
+			counterparty := types.NewCounterparty(clientA, connA.ID, suite.chainA.GetPrefix())
+
+			connectionKey := host.KeyConnection(connA.ID)
+			proofInit, proofHeight := suite.chainA.QueryProof(connectionKey)
 
-			connectionKey := host.KeyConnection(testConnectionIDA)
-			proofInit, proofHeight := queryProof(suite.chainA, connectionKey)
+			// retrieve consensus state to provide proof for
```

Comment:
added making this a helper as a todo in the ibc testing issue, noticed myself repeating it